### DISCUSSION TITLE
How can I store plane_normal_vector and direction_vector in crystal plasticity framework?

### DISCUSSION THREAD

**Discussion by LifengGan** [idaholab/moose#26088] [Q&A Modules: Solid mechanics] [answered]
Dear moose developers,

I want to store `plane_normal_vector` (m) and `direction_vector` (n) in crystal plasticity framework, and I take the reference from @ngrilli, (see the link below), through the reclaim the `calculateSchmidTensor()` again.

And I did the same thing, but it will never go through this  `calculateSchmidTensor()`. I got all the value were the initialized one, i.e., zero.

Since in file `CrsytalPlasticityStressUpdateBase.C`, it already has one `calculateSchmidTensor()`, so I want to ask about the logic by recaiming this functoin.

https://github.com/ngrilli/c_pfor_am/blob/HEAD/src/materials/CrystalPlasticityDislocationUpdate.C#L251

Thanks.

LifengGan

**Comment by GiudGiud** [comment, marked as answer]
Hello

This is a virtual function so if you want to change its definition you should 
- make a derived class from this CrystalPlasticityUpdate(Base) class
- override it by redefining the function. Don't forget the 'override' keyword at the end of the function declaration (in the header), it will make sure the override succeeded

Guillaume

**Reply by LifengGan** [reply]
Hi,

However, the reference in header file 

https://github.com/ngrilli/c_pfor_am/blob/main/include/materials/CrystalPlasticityDislocationUpdate.h#L44

there is no override keywork, and if i add the override keywork, get the `marked "override", but does not override` error when make file

And I found in `CrystalPlasticityUpdateBase` class in moose, the `calculateSchmidTensor()` is not a virtual function, I cannot override this function.

The question back to the origin. How can I store` plane_normal_vector` and `direction_vector `in crystal plasticity framework?

**Reply by LifengGan** [reply]
I did solve the problem by copy and modify in both files ` ComputeMultipleCrystalPlasticityStress` and `CrystalPlasiticyUpdateBase` for modifying `calculateSchmidTensor()` to the virtual one.

I just wondering if there a simple way to solve thi problem.

Thanks
